### DISCUSSION TITLE
Can I customize useRouter so that I can use both the app directory and the page directory?

### DISCUSSION THREAD

**Discussion by nerdchanii** [vercel/next.js#52829] [Help] [answered]
### Summary

I'm trying to create a library to provide compatibility for both the app directory and the page directory. The library is using 'next/router', and the error 'NextRouter was not mounted' occurs when using it in the app directory. (I checked it in the document.)

Can "useRouter" in "Next/Router" and "Next/Navigation" be used with strategy patterns, etc.? The objects and types of the two routers are different, but to some extent, they seem to be compatible.

### Additional information

```js
import {useRouter as useNextRouter} from 'next/router';
import {useRouter as useAppRouter} from 'Next/Navigation';

const Custom Router = ()=>{
  try{
    return useNextRouter();
} Catch(e){
   return useAppRouter();
}



Is it okay to use it like this?
```

**Comment by icyJoseph** [comment, marked as answer]
No, I don't think that's even a good idea in principle, because the router API is vastly different, mostly that the App Dir Router has only navigation methods, and no readable properties. Those are handled through other context providers.

I think this would also crash as soon as the app dir module parser finds `next/router` imported. Perhaps this could still be hacked, but it is just asking for trouble.

This sounds like a good use case for dependency injection. Your framework/library should take in a router instance, do type checking on it, whether it is app router or pages router, and use them accordingly.

**Reply by nerdchanii** [reply]
Oh, Thank you for bringing that to my attention. I missed the importance of implementing dependency injection.

You made a point about this situation being a good example of why dependency injection is beneficial. Thanks to your answer, I now realize that the approach I mentioned earlier isn't the right one. I appreciate your guidance. Thank you.
